### ISSUE TITLE
Magento 2.1 does not allow adding product images after a product is created

### ISSUE THREAD

**Issue by blizzardengle** [magento/magento2#5845]
Magento 2.1 will only allow me to upload pictures of a product when I initially create them. If I make a new product and choose to leave out pictures I can not go back in to add pictures later. If I edit the product and attempt to add images I get the following error:

_**Attention: We don't recognize or support this file extension type.**_

This was tested with simple product and configurable products. This is not the same issue as ticket [#5746](https://github.com/magento/magento2/issues/5746) because no errors are reported by Magento besides the above warning on the product page and my Magento install is 100% complete and installed correctly.
### Preconditions
1. Magento 2.1 not Magento 2.0.X
2. Linux or Windows servers; I have tested both including on different PHP versions 5.6 to 7.0
### Steps to reproduce
1. Create a new simple or configurable product in Magento 2.1 and do not add product images.
2. Save and leave the product; again do not add any images.
3. Navigate around the dashboard for a bit and then go back and edit the product you created.
4. Attempt to add images now and you should see the error. 
### Expected result

You should be able to add, edit, and delete product images at any time not just when you first create the product.
### Actual result

Once a product is created you can never add images to it unless you added some when you created the product. You also get this error every time you try:

_**Attention: We don't recognize or support this file extension type.**_

**Comment by andidhouse** [comment]
Can not reproduce this on a 2.1 with php7. 

--> Create simple product and save without image
--> Navigate around in the backend
--> Go back to product und upload and save pictures works

**Comment by blizzardengle** [comment]
Try PHP 5.6. 

I did a fresh install on CentOS and on my local Windows machine using Magento 2.1 and PHP 5.6 and it happened both times. This could be caused by something else but I have found nothing in the logs to warn me that some other error is happening.

**Comment by blizzardengle** [comment]
Update on my progress trying to troubleshoot this bug.

**Update**<br>
XHR response is: 

{"error":"The file was not uploaded.","errorcode":666}

I double checked that upload_tmp_dir was set correctly and still no change. Images are not uploaded and this error is always returned by the ajax request.

**Comment by Cacasapo** [comment]
I'm on CentOS 7, PHP 5.6.24 and can't reproduce the issue. Fresh install of 2.1 I'm using for testing.
Sounds like a permissions issue to me.

**Comment by blizzardengle** [comment]
**Solved**<br>
This error can be caused by several server configuration issues. My issue was with Insufficient max file size and post file size. Each error will always present itself with the same generic warning from Magento: **_Attention: We don't recognize or support this file extension type.**_ Below are the three possible server settings that are causing this error. There may be more than this list but these are all I have personally tested. **Magento should really add some of this to the documentation or at least the warning message that is displayed.**

**1) Permission Problem**<br>
You may get an error along the lines of "Can not create directory pub/media/tmp" or you may notice after examining the XHR summary in developer tools that a 200 success code is returned but no data was saved to the server. The easiest way to test this to check your servers permissions and make sure all the directories, especially everything under pub/media is 755. You may have one of the next two issues as well so do not keep troubleshooting permissions until you check them as well.

**2) Missing Write Permissions to Upload Directory / Missing Directory**<br>
PHP needs to use the servers temporary upload directory when uploading files to the server. Make sure that your server has a temporary upload directory setup and is writable by the Magento user; whichever user you are using for your Magento website. Most hosting providers have this setup for you but some may lock down this feature or may be missing a default setting. Private servers sometimes rely on you choosing the temporary upload directory per domain and then defaulting to a generic location if you failed to choose one. Look in your php.ini (phpinfo) for upload_tmp_dir. No value means the server will use the default if any.

**3) Insufficient max file size and post file size**<br>
In your php.ini file you need to correctly set your upload_max_filesize and  post_max_size. If any image or any upload of images exceeds these settings you will get the error from the OP. Remember that post_max_size must be equal to or greater than upload_max_filesize. Also note that your web host will most likely not warn you if you entered an incorrect value here. You could mistype the value or choose a size the web host does not allow. Test for this.

**Test Script**<br>
The following is a PHP script you can use to test for some of these issues. Make a simple HTML form with one file input named upfile and point the form submission to this script. You will need to create an uploads directory at the came location that you run this script. It is not my code, it was originally posted to the [PHP Manual](http://php.net/manual/en/features.file-upload.php) by CertaiN.

```
<?php

header('Content-Type: text/plain; charset=utf-8');

try {

// Undefined | Multiple Files | $_FILES Corruption Attack
// If this request falls under any of them, treat it invalid.
if (
    !isset($_FILES['upfile']['error']) ||
    is_array($_FILES['upfile']['error'])
) {
    throw new RuntimeException('Invalid parameters.');
}

// Check $_FILES['upfile']['error'] value.
switch ($_FILES['upfile']['error']) {
    case UPLOAD_ERR_OK:
        break;
    case UPLOAD_ERR_NO_FILE:
        throw new RuntimeException('No file sent.');
    case UPLOAD_ERR_INI_SIZE:
    case UPLOAD_ERR_FORM_SIZE:
        throw new RuntimeException('Exceeded filesize limit.');
    default:
        throw new RuntimeException('Unknown errors.');
}

// You should also check filesize here. 
if ($_FILES['upfile']['size'] > 1000000) {
    throw new RuntimeException('Exceeded filesize limit.');
}

// DO NOT TRUST $_FILES['upfile']['mime'] VALUE !!
// Check MIME Type by yourself.
$finfo = new finfo(FILEINFO_MIME_TYPE);
if (false === $ext = array_search(
    $finfo->file($_FILES['upfile']['tmp_name']),
    array(
        'jpg' => 'image/jpeg',
        'png' => 'image/png',
        'gif' => 'image/gif',
    ),
    true
)) {
    throw new RuntimeException('Invalid file format.');
}

// You should name it uniquely.
// DO NOT USE $_FILES['upfile']['name'] WITHOUT ANY VALIDATION !!
// On this example, obtain safe unique name from its binary data.
if (!move_uploaded_file(
    $_FILES['upfile']['tmp_name'],
    sprintf('./uploads/%s.%s',
        sha1_file($_FILES['upfile']['tmp_name']),
        $ext
    )
)) {
    throw new RuntimeException('Failed to move uploaded file.');
}

echo 'File is uploaded successfully.';

} catch (RuntimeException $e) {

echo $e->getMessage();

}

?>
```

**Comment by magento-engcom-team** [comment]
@blizzardengle, thank you for your report.
We were not able to reproduce this issue by following the steps you provided. If you'd like to update it, please reopen the issue.

**Comment by rip057** [comment]
for those that may experience this issue, what cured this issue for me was to clear the /pub/media/tmp/ directory out.  it seems that older versions permissions were set on old temp directories.  obviously make back ups, but clearing the tmp directory allowed me to fix this.

**Comment by saeohioalpha** [comment]
I have the same issue. I ran the script above and get the file uploaded successfully. I also have all the permissions set. I recently updated from 2.1.8 to 2.1.12 and the issue persisted on both.

From what I can tell, the issue is, in my case, that I have a redirect for ?SID in my NGINX config, because magento annoyingly keeps adding it even when disabled in the URL and it keeps popping up on my indexing in google search console. Well, the /upload would actually be /upload/?SID=!@#. Which would get a 301 to just /upload/ this caused the image upload to report this error.

The odd thing is I had this redirect in since may 2017, and have been at 2.1.8 since probably late aug 2017, but the issue just started within the last few weeks or so of march 2018.

**Comment by blizzardengle** [comment]
For people who keep stumbling on this issue I would recommend looking at [my answer on the Magento Stack Exchange website](https://magento.stackexchange.com/a/130229/28324). Most issue can be resolved by following those troubleshooting steps in Magento 2.1+ and Magento 2.2+ where a few new issue popped up now that PHP 7 is required.

@saeohioalpha I don't think my post covers your issue unfortunately since I've never used Magento on Nginx but I would love to add the solution to it when you figure it out.

**Comment by saeohioalpha** [comment]
Thanks, that post is where I started my troubleshooting. I had covered all those bases, and still nothing. Like I said the issue somehow was that the session id was still being added to the URL for the image upload. Not sure why, but once I removed the 301 redirect for /xxxx?SID to /xxxx the uploads started working. I will revisit once I upgrade to M2.2 and see if it still persists. 

It is just odd that the issue manifested just recently when that redirect had been in for over 6 mos.
